src/loop.cpp

```diff
@@ -29,6 +29,111 @@ namespace oxen::quic
         });
     }
 
+    /** Static casting to `decltype(timeval::tv_{sec,usec})` makes sure that;
+        - on linux
+            .tv_sec is type __time_t
+            .tv_usec is type __suseconds_t
+        - on OSX    (https://developer.apple.com/documentation/kernel/timeval)
+            .tv_sec is type __darwin_time_t
+                - this is an annoying typedef of `time_t`
+            .tv_usec is type __darwin_suseconds_t
+                - this is an equally annoying typedef for `suseconds_t`
+        Alas, yet again another mac idiosyncrasy...
+     */
+    timeval loop_time_to_timeval(loop_time t)
+    {
+        return timeval{
+                .tv_sec = static_cast<decltype(timeval::tv_sec)>(t / 1s),
+                .tv_usec = static_cast<decltype(timeval::tv_usec)>((t % 1s) / 1us)};
+    }
```

Comment:
I don't understand the comment/code: on both platforms it appears that `tv_sec` is a `time_t` and `tv_usec` is a `suseconds_t` (regardless of the intermediate internal typedef name), so it ought to suffice to `static_cast<time_t>` and `static_cast<suseconds_t>` here.

Reply:
I explicitly specified this to fix an issue in CI, you can change it back and see how it will fail the mac builds

Reply:
There are tons of issues online documenting [this](https://github.com/szymonwieloch/rust-rawsock/issues/5)

Reply:
macos seems happy with this:
```bash
om2g:~$ cat timeval.cpp 
```
```C++
#include <sys/time.h>
#include <chrono>
#include <iostream>

using namespace std::literals;

timeval loop_time_to_timeval(std::chrono::microseconds t)
{
    return timeval{
            .tv_sec = static_cast<time_t>(t / 1s),
            .tv_usec = static_cast<suseconds_t>((t % 1s) / 1us)};
}

int main() {
    auto x = loop_time_to_timeval(123456ms);
    std::cout << x.tv_sec << "s, " << x.tv_usec << "µs\n";
}
```
```bash
om2g:~$ g++ -std=c++20 timeval.cpp -o timeval
om2g:~$ ./timeval 
123s, 456000µs
```

Do you have a CI job number where this was failing?

Reply:
Can you explain why that is an improvement over the current cast to whatever the system defined `timeval` struct is? The above seems insufficient for other system implementations that can [compile things away](https://codebrowser.dev/gcc/include/x86_64-linux-gnu/bits/types/struct_timeval.h.html), like 64-bit `__time64_t`

Reply:
Because `timeval` is a [documented, decades old POSIX standard](https://pubs.opengroup.org/onlinepubs/7908799/xsh/systime.h.html) that is required to have a `time_t tv_sec;` and `suseconds_t tv_usec;`.  Whatever `__time64_t` or `__darwin_suseconds_t` get used by standard library implementations is irrelevant: those are internal implementation types which are required to be the same as `time_t` and `suseconds_t` typedef in order to satify the `timeval` specification.

Reply:
Except `time_t` is *itself* a [typedef](https://en.wikipedia.org/w/index.php?title=Time_t&oldid=450752800):

>This type is a [typedef](https://en.wikipedia.org/wiki/Typedef) defined in the standard [<time.h>](https://en.wikipedia.org/wiki/Time.h) header. [ISO C](https://en.wikipedia.org/wiki/ISO_C) defines time_t as an arithmetic type, but does not specify any particular [type](https://en.wikipedia.org/wiki/Data_type), [range](https://en.wikipedia.org/wiki/Range_(mathematics)), [resolution](https://en.wikipedia.org/wiki/Frequency), or encoding for it. Also unspecified are the meanings of arithmetic operations applied to time values.

Casting to the native timeval type is the straight-line, most efficient solution to this.

Reply:
> Except time_t is itself a typedef

Yup.  It's a typedef that is explicitly defined to be the type that goes into a `tv_sec`, and `suseconds_t` is likewise a typedef that is defined to be the type that goes into `tv_usec`.

Reply:
That is also incorrect. Please feel free to [verify this yourself](https://en.cppreference.com/w/c/chrono/time_t). For extra thoroughness, please see the latest [C17 standard](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf), specifically section 7.27.1-4 to see:

> The range and precision of times representable in clock_t and time_t are implementation-defined

`time_t` is not typedef'ed specifically for `tv_sec`. It is used in a whole host of other functions that can be enumerated by searching "time_t" on the C17 standard pdf file.

On my system, `time_t` is a typedef for `__time_t`, which *itself* typedefs a few times all the way down to `long int` on my system

Reply:
That is also incorrect, it is not typedef'ed specifically for that use-case. Please feel free to [verify this yourself](https://en.cppreference.com/w/c/chrono/time_t)
> Although not defined by the C standard, this is almost always an integral value holding the number of seconds

For reference to what they're speaking about, please see the latest [C17 standard](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf), specifically section 7.27.1-4 to see:

> The range and precision of times representable in clock_t and time_t are implementation-defined

`time_t` is not typedef'ed specifically for `tv_sec`. It is used in a whole host of other functions that can be enumerated by searching "time_t" on the C17 standard pdf file.

On my system, `time_t` is a typedef for `__time_t`, which *itself* typedefs a few times all the way down to `long int` on my system

Reply:
Not resolved.

Again: `timeval` is explicitly defined in terms of `time_t` and `suseconds_t`, and thus *those are the types* that belong here.  Not `__time_t`, not `decltype(...)`, not `long int`, not `__darwin_time_t`, but simply the member types that the timeval API explicitly defines, which are `time_t` and `suseconds_t`.

Reply:
Well hasn't this certainly been an educational experience. In the interest of not continuing to block this for a third week over named booleans and other nit-picks, it is changed to read as you want it

Reply:
[Beautiful](https://ci.oxen.rocks/oxen-io/oxen-libquic/1115/23/3)

Reply:
[Well](https://ci.oxen.rocks/oxen-io/oxen-libquic/1116/23/3), look like we can securely resolve this now.